### PR TITLE
Test nfts async wrong args

### PR DESCRIPTION
**Pull request by sasurobert**
vmcommon integration and test

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -12,7 +12,7 @@ require (
 	github.com/ElrondNetwork/elrond-go-core v1.1.16-0.20220708085217-ccc0c5ac9076
 	github.com/ElrondNetwork/elrond-go-crypto v1.0.1
 	github.com/ElrondNetwork/elrond-go-logger v1.0.7
-	github.com/ElrondNetwork/elrond-vm-common v1.3.13
+	github.com/ElrondNetwork/elrond-vm-common v1.3.14-0.20220715122033-be46f4a725c5
```
Comment (iulianpascalau) [inline go.mod:15]: let's not forget about proper releases

File: integrationTests/vm/testInitializer.go [hunk 2]
```diff
@@ -563,7 +564,7 @@ func CreateVMAndBlockchainHookAndDataPool(
 		AutomaticCrawlerAddress:   bytes.Repeat([]byte{1}, 32),
 		MaxNumNodesInTransferRole: 100,
 	}
-	builtInFuncs, nftStorageHandler, globalSettingsHandler, _ := builtInFunctions.CreateBuiltInFuncContainerAndNFTStorageHandler(argsBuiltIn)
+	builtInFuncFactory, _ := builtInFunctions.CreateBuiltInFunctionsFactory(argsBuiltIn)
```
Comment (iulianpascalau) [inline integrationTests/vm/testInitializer.go:567]: naming: this is rather a handler, not a factory :D

File: integrationTests/vm/txsFee/multiShard/asyncCall_test.go [hunk 3]
```diff
@@ -122,8 +122,8 @@ func TestAsyncCallShouldWork(t *testing.T) {
 	scr = intermediateTxs[0]
 	utils.ProcessSCRResult(t, testContextSecondContract, scr, vmcommon.Ok, nil)
 
-	require.Equal(t, big.NewInt(49993150), testContextSecondContract.TxFeeHandler.GetAccumulatedFees())
-	require.Equal(t, big.NewInt(4999315), testContextSecondContract.TxFeeHandler.GetDeveloperFees())
+	require.Equal(t, big.NewInt(3680), testContextSecondContract.TxFeeHandler.GetAccumulatedFees())
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/multiShard/asyncCall_test.go:125]: why these changes?
If these changes are required because the activation flag is set for the transferAndAsyncCallbackFixEnableEpoch, then, I think we should double the tests, one with the active flag and one with the disabled flag in order to easily assure backwards compatibility.

File: integrationTests/vm/txsFee/multiShard/asyncESDT_test.go [hunk 4]
```diff
@@ -102,10 +102,10 @@ func TestAsyncESDTTransferWithSCCallShouldWork(t *testing.T) {
 	_, err = testContextSender.Accounts.Commit()
 	require.Nil(t, err)
 
-	expectedAccumulatedFees = big.NewInt(189880)
+	expectedAccumulatedFees = big.NewInt(189890)
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/multiShard/asyncESDT_test.go:105]: here also. What changed in terms of fees?
If these changes are required because the activation flag is set for the transferAndAsyncCallbackFixEnableEpoch, then, I think we should double the tests, one with the active flag and one with the disabled flag in order to easily assure backwards compatibility.

File: process/coordinator/transactionType.go [hunk 5]
```diff
@@ -2,6 +2,7 @@ package coordinator
 
 import (
 	"bytes"
+	"github.com/ElrondNetwork/elrond-go-core/core/atomic"
```
Comment (iulianpascalau) [inline process/coordinator/transactionType.go:5]: go imports

File: process/coordinator/transactionType_test.go [hunk 6]
```diff
@@ -3,6 +3,7 @@ package coordinator
 import (
 	"bytes"
 	"encoding/hex"
+	"github.com/ElrondNetwork/elrond-go/testscommon/epochNotifier"
```
Comment (iulianpascalau) [inline process/coordinator/transactionType_test.go:6]: go imports

File: process/smartContract/process.go [hunk 7]
```diff
@@ -300,6 +304,9 @@ func (sc *scProcessor) GasScheduleChange(gasSchedule map[string]map[string]uint6
 	}
 
 	sc.builtInGasCosts = builtInFuncCost
+	if sc.flagFixAsyncCallBackArgumentsParser.IsSet() {
+		sc.builtInGasCosts[core.BuiltInFunctionMultiESDTNFTTransfer] = builtInFuncCost["ESDTNFTMultiTransfer"]
```
Comment (iulianpascalau) [inline process/smartContract/process.go:308]: this will be pretty hard to keep track. Why not create a new gas map that will be activated exactly when the flagFixAsyncCallBackArgumentsParser is set?

File: process/smartContract/process.go [hunk 8]
```diff
@@ -1234,13 +1246,20 @@ func (sc *scProcessor) createVMInputWithAsyncCallBackAfterBuiltIn(
...
-		function, args, err := sc.argsParser.ParseCallData(string(outAcc.OutputTransfers[0].Data))
-		log.LogIfError(err, "function", "createVMInputWithAsyncCallBackAfterBuiltIn.ParseCallData")
-		if len(function) > 0 {
-			arguments = append(arguments, []byte(function))
-		}
 
-		arguments = append(arguments, args...)
+		if sc.flagFixAsyncCallBackArgumentsParser.IsSet() {
+			args, err := sc.argsParser.ParseArguments(string(outAcc.OutputTransfers[0].Data))
+			log.LogIfError(err, "function", "createVMInputWithAsyncCallBackAfterBuiltIn.ParseArguments")
```
Comment (iulianpascalau) [inline process/smartContract/process.go:1252]: why not returning error here? Otherwise, the code execution will simply append a nil args slice and jump to L1265

File: process/smartContract/process.go [hunk 9]
```diff
@@ -1957,6 +1976,9 @@ func (sc *scProcessor) penalizeUserIfNeeded(
 	if callType == vmData.AsynchronousCall {
 		return
 	}
+	if callType == vmData.AsynchronousCallBack {
```
Comment (iulianpascalau) [inline process/smartContract/process.go:1979]: I think we need a flag for this

File: integrationTests/vm/txsFee/multiShard/asyncESDT_test.go (hunk 4 above)
Comment (sasurobert) [reply integrationTests/vm/txsFee/multiShard/asyncESDT_test.go:105]: because the consumption was wrong. it was completely wrong for different reasons - like 2 fixes in process.go

File: integrationTests/vm/txsFee/multiShard/asyncCall_test.go (hunk 3 above)
Comment (sasurobert) [reply integrationTests/vm/txsFee/multiShard/asyncCall_test.go:125]: because the consumption was wrong. it was completely wrong for different reasons - like 2 fixes in process.go. Backward compatibility is ensured by flags.

File: process/coordinator/transactionType.go (hunk 5 above)
Comment (sasurobert) [reply process/coordinator/transactionType.go:5]: done

File: process/coordinator/transactionType_test.go (hunk 6 above)
Comment (sasurobert) [reply process/coordinator/transactionType_test.go:6]: done

File: process/smartContract/process.go (hunk 7 above)
Comment (sasurobert) [reply process/smartContract/process.go:308]: Here we have an issue of naming in gasSchedule vs builtInFunctions name and sc.bultInGasCosts[function] returned 0. Which was wrong. This is the case of different string constants between core.BuiltInFunctionMultiESDTNFTTransfer and the one from gas schedule. If I add it into the gasSchedule it will not be backward compatible.

File: process/smartContract/process.go (hunk 8 above)
Comment (sasurobert) [reply process/smartContract/process.go:1252]: would leave as it is - as it calls back the contract (asyncCallBack) with empty args and that's it.

File: integrationTests/vm/txsFee/multiShard/asyncCall_test.go (hunk 3 above)
Comment (sasurobert) [reply integrationTests/vm/txsFee/multiShard/asyncCall_test.go:125]: We can duplicate one of these tests if needed.

File: integrationTests/vm/testInitializer.go (hunk 2 above)
Comment (sasurobert) [reply integrationTests/vm/testInitializer.go:567]: FactoryHandler to be precise 📦

File: process/smartContract/process.go (hunk 7 above)
Comment (iulianpascalau) [reply process/smartContract/process.go:308]: It will be, only in a v7 of the gas map that will be active in the same epoch with the fix

File: integrationTests/vm/testInitializer.go (hunk 2 above)
Comment (iulianpascalau) [reply integrationTests/vm/testInitializer.go:567]: no, handler, as it stores components. Factory just creates pointers but do not hold them

File: integrationTests/vm/txsFee/multiShard/asyncCall_test.go (hunk 3 above)
Comment (iulianpascalau) [reply integrationTests/vm/txsFee/multiShard/asyncCall_test.go:125]: ok, got it

File: process/smartContract/process.go (hunk 7 above)
Comment (iulianpascalau) [reply process/smartContract/process.go:308]: Understood the problem, I do not have a better way to fix this since all variants will involve a flag like this

**Review by iulianpascalau — DISMISSED** [review]

**Review by andreibancioiu — APPROVED** [review]
Tests passed.

**Review by popenta — DISMISSED** [review]
@@                            Log scanner                                  @@
##                            test-nfts-async-wrong-args                   ##
================================================================================
- Known Warnings              16
- New Warnings                5
- Known Errors                0
- New Errors                  2
- Panics                      0
================================================================================

**Review by AdoAdoAdo — APPROVED** [review]

File: integrationTests/vm/txsFee/multiShard/asyncESDT_test.go [hunk 10]
```diff
@@ -136,7 +136,7 @@ func TestAsyncESDTTransferWithSCCallShouldWork(t *testing.T) {
 
 	utils.ProcessSCRResult(t, testContextFirstContract, intermediateTxs[1], vmcommon.Ok, nil)
 
-	require.Equal(t, big.NewInt(278300), testContextFirstContract.TxFeeHandler.GetAccumulatedFees())
+	require.Equal(t, big.NewInt(4936720), testContextFirstContract.TxFeeHandler.GetAccumulatedFees())
```
Comment (AdoAdoAdo) [inline integrationTests/vm/txsFee/multiShard/asyncESDT_test.go:139]: There is quite a large change on the expected values.
how can we check that this value is correct?

Comment (sasurobert) [reply integrationTests/vm/txsFee/multiShard/asyncESDT_test.go:139]: checked manually the code and we get the new values. It was a problem on builtInFunction costs.

**Review by popenta — APPROVED** [review]
system test passed

**Review by andreibancioiu — APPROVED** [review]
